lib/providers/update_provider.dart

```diff
@@ -0,0 +1,372 @@
+    required this.publishedAt,
+  });
+
+  @override
+  String toString() =>
+      'UpdateVersion(version: $version, description: $description, publishedAt: $publishedAt)';
+
+  @override
+  bool operator ==(Object other) {
+    if (identical(this, other)) return true;
+
+    return other is UpdateVersion &&
+        other.version == version &&
+        other.description == description &&
+        other.publishedAt == publishedAt;
+  }
+
+  @override
+  int get hashCode =>
+      version.hashCode ^ description.hashCode ^ publishedAt.hashCode;
+
+  static const titleField = 'title';
+  static const descriptionField = 'description';
+  static const publishedAtField = 'pubDate';
+
+  static const rpm = 'rpm';
+  static const deb = 'deb';
+  static const tarball = 'tar.gz';
+  static const appImage = 'appimage';
+  static const linuxDownloadFileName = 'bluecherry-linux-x86_64';
+  static const windowsDownloadFileName = 'bluecherry-windows-setup';
+}
+
+class UpdateManager extends ChangeNotifier {
+  UpdateManager._();
+
+  /// `late` initialized [UpdateManager] instance.
+  static late final UpdateManager instance;
+  late final PackageInfo packageInfo;
+  late String tempDir;
+
+  /// The URL to the appcast file.
+  static const appCastUrl =
+      'https://raw.githubusercontent.com/bdlukaa/unity/upgrade/bluecherry_appcast.xml';
+
+  /// Initializes the [UpdateManager] instance & fetches state from `async`
+  /// `package:hive` method-calls. Called before [runApp].
+  static Future<UpdateManager> ensureInitialized() async {
+    instance = UpdateManager._();
+    await instance.initialize();
+    return instance;
+  }
+
+  /// If true, there is a new update available.
+  ///
+  /// If false, the user is up to date with the latest version.
+  bool get hasUpdateAvailable {
+    final currentVersion = packageInfo.version;
+    return Version.parse(currentVersion) !=
+        Version.parse(latestVersion.version);
+  }
+
+  List<UpdateVersion> versions = [];
+  UpdateVersion get latestVersion => versions.last;
+
+  Future<void> initialize() async {
+    final data = await downloads.read() as Map;
+    if (!data.containsKey(kHiveAutomaticUpdates)) {
+      await _save();
+    } else {
+      await _restore();
+    }
+
+    await Future.wait([
+      checkForUpdates(),
+      PackageInfo.fromPlatform().then((result) {
+        packageInfo = result;
+      })
+    ]);
+
+    if (hasUpdateAvailable && automaticDownloads) {
+      download(latestVersion.version);
+    }
+  }
+
+  Future<void> _save({bool notify = true}) async {
+    await downloads.write({
+      kHiveAutomaticUpdates: automaticDownloads,
+      kHiveLastCheck: lastCheck?.toIso8601String(),
+    });
+
+    if (notify) notifyListeners();
+  }
+
+  Future<void> _restore({bool notifyListeners = true}) async {
+    final data = await downloads.read() as Map;
+
+    _automaticDownloads = data[kHiveAutomaticUpdates];
+    _lastCheck = DateTime.tryParse(data[kHiveLastCheck] ?? '');
```

Comment:
I'd rather check if last check is null than making `?? ''`, which can stress the code out